### ISSUE TITLE
Annotate CSharpSyntaxVisitor.Visit as [return: NotNullIfNotNull("node")]

### ISSUE THREAD

**Issue by jnm2** [dotnet/roslyn#42436]
I noticed a lot of places in my code where I accidentally pass a nullable node to Visit and cast the return to a non-nullable value even though null might be returned.

In order to get a warning about this, please enable nullability annotations for this class and add these attributes:

```cs
[return: MaybeNull, NotNullIfNotNull("node")]
```

Current source:

https://github.com/dotnet/roslyn/blob/e4c5a8c88eee6358c914a51adfd0fab87d1c79fc/src/Compilers/CSharp/Portable/Syntax/CSharpSyntaxVisitor.cs#L22-L31

**Comment by jnm2** [comment]
Workaround for now:

```cs
[return: MaybeNull, NotNullIfNotNull("node")]
public new VisualBasicSyntaxNode Visit(SyntaxNode? node) => base.Visit(node);
```

**Comment by jnm2** [comment]
CSharpSyntaxRewriter also needs this.

**Comment by CyrusNajmabadi** [comment]
I don't think we can actually do this.  It's ultimately going to depend on the behavior of hte visitor.  It may indeed return 'null' in visit methods, regardless of hte argument.  For example, in the rewriter, 'null' is legal for a visit method to return.  this can be totally ok.  For example, considering visiting `return expr;` and returning `null` when visiting the expr here.  That would legally create `return;` as the expr child is ok to be null.

Similarly, for any node in a list, it's ok to return null back, and that null value will simply not be placed into the new list being built up. 

I genuinely think the contract is that it is expected to get null back here.
